cirq/protocols/resolve_parameters.py

```diff
@@ -67,6 +79,68 @@ def is_parameterized(val: Any) -> bool:
         return False
 
 
+def parameter_names(val: Any, *,
+                    check_symbols: bool = True) -> AbstractSet[str]:
+    """Returns parameter names for this object.
+
+    Args:
+        val: Object for which to find the parameter names.
+        check_symbols: If true, fall back to calling parameter_symbols.
+
+    Returns:
+        A set of parameter names if the object is parameterized. It the object
+        does not implement the _parameter_names_ magic method or that method
+        returns NotImplemented, returns an empty set.
+    """
+    if isinstance(val, sympy.Basic):
+        return {symbol.name for symbol in val.free_symbols}
+    if isinstance(val, (list, tuple)):
+        return {name for e in val for name in parameter_names(e)}
+
+    getter = getattr(val, '_parameter_names_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is not NotImplemented:
+        return result
+
+    if check_symbols:
+        symbols = parameter_symbols(val, check_names=False)
+        if symbols is not NotImplemented:
+            return {symbol.name for symbol in symbols}
+
+    return set()
+
+
+def parameter_symbols(val: Any, *,
+                      check_names: bool = True) -> AbstractSet[sympy.Symbol]:
+    """Returns parameter symbols for this object.
+
+    Args:
+        val: Object for which to find the parameter symbols.
+        check_names: If true, fall back to calling parameter_names.
+
+    Returns:
+        A set of parameter symbols if the object is parameterized. It the object
+        does not implement the _parameter_symbols_ magic method or that method
+        returns NotImplemented, returns an empty set.
+    """
+    if isinstance(val, sympy.Basic):
+        return val.free_symbols
+    if isinstance(val, (list, tuple)):
+        return {symbol for e in val for symbol in parameter_symbols(e)}
+
+    getter = getattr(val, '_parameter_symbols_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is not NotImplemented:
+        return result
+
+    if check_names:
+        names = parameter_names(val, check_symbols=False)
+        if names is not NotImplemented:
+            return {sympy.Symbol(name) for name in names}
+
+    return set()
```

Comment:
The code of these two functions is nearly identical. Do we really need both? At minimum I'd implement one of these in terms of the other, but maybe it's even better to leave the conversion up to the user. I imagine any client code can easily be adapted to work with either.

Reply:
I'd lean toward implementing `parameter_symbols` in terms of `parameter_names`, since strings are built-in. But that would incur extra overhead in some cases, e.g. if getting symbols from a sympy expression we would convert to strings and then back to symbols. Not a big deal, but something to consider.

We have a somewhat unfortunate dichotomy between strings and symbols throughout the parametrization code where strings and symbols are in 1:1 correspondence but one or the other is more useful in different cases. I'd rather not force users who want to work with symbols to always have to convert from strings (`{sympy.Symbol(n) for n in cirq.parameter_names(val)}`) and vice-versa, which is why I thought to add two functions.